src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs

```diff
@@ -39,18 +40,87 @@ public override bool TryRead(Type valueType, ref Utf8JsonReader reader, out TVal
                 return false;
             }
 
-            if (s_isUint64)
-            {
-                if (reader.TryGetUInt64(out ulong ulongValue))
-                {
-                    value = (TValue)Enum.ToObject(valueType, ulongValue);
-                    return true;
-                }
-            }
-            else if (reader.TryGetInt64(out long longValue))
+            // When utf8reader/writer will support all primitive types we should remove custom bound checks
+            // https://github.com/dotnet/corefx/issues/36125
+
+            switch (s_enumTypeCode)
             {
-                value = (TValue)Enum.ToObject(valueType, longValue);
-                return true;
+                case TypeCode.SByte:
+                    {
+                        if (reader.TryGetInt32(out int byte8) && JsonHelpers.IsInRangeInclusive(byte8, sbyte.MinValue, sbyte.MaxValue))
+                        {
+                            sbyte byte8Value = (sbyte)byte8;
+                            value = Unsafe.As<sbyte, TValue>(ref byte8Value);
+                            return true;
+                        }
+                        break;
+                    }
+                case TypeCode.Byte:
+                    {
+                        if (reader.TryGetUInt32(out uint ubyte8) && JsonHelpers.IsInRangeInclusive(ubyte8, byte.MinValue, byte.MaxValue))
+                        {
+                            byte ubyte8Value = (byte)ubyte8;
+                            value = Unsafe.As<byte, TValue>(ref ubyte8Value);
+                            return true;
+                        }
+                        break;
+                    }
+                case TypeCode.Int16:
+                    {
+                        if (reader.TryGetInt32(out int int16) && JsonHelpers.IsInRangeInclusive(int16, short.MinValue, short.MaxValue))
+                        {
+                            short shortValue = (short)int16;
+                            value = Unsafe.As<short, TValue>(ref shortValue);
+                            return true;
+                        }
+                        break;
+                    }
+                case TypeCode.UInt16:
+                    {
+                        if (reader.TryGetUInt32(out uint uint16) && JsonHelpers.IsInRangeInclusive(uint16, ushort.MinValue, ushort.MaxValue))
+                        {
+                            ushort ushortValue = (ushort)uint16;
+                            value = Unsafe.As<ushort, TValue>(ref ushortValue);
+                            return true;
+                        }
+                        break;
+                    }
+                case TypeCode.Int32:
```

Comment:
Nit: these should be ordered by how often they are expected to occur. Given that int is the default, it should come first. My guess is that after that it would be unsigned types and lastly signed.

You don't have to do that in this change- I'm working on an update for this to support strings so I can take care of ordering.

Reply:
Ok let me know if I need to do other on this!Thank's for review.